### PR TITLE
Streamline Grafana/Influx/Prometheus Docker config

### LINKED ISSUE #2900: Streamline Grafana/Influx/Prometheus Docker configuration support
- [x] grafana image to be made from src/grafana/prometheus/grafana
 -- lives in spring-cloud DH org
 -- name: `springcloud/spring-cloud-dataflow-grafana-prometheus`
 -- image version would be tied to DF version (reuse the`DATAFLOW_VERSION` variable)
- [x]  grafana image to be made from src/grafana/influxdb/grafana
 -- lives in spring-cloud DH org
 -- name: `springcloud/spring-cloud-dataflow-grafana-influxdb`
 -- image version would be tied to DF version (reuse the`DATAFLOW_VERSION` variable)
- [x]  The `src/grafana` docker-compose files should use the public image vs. in-tree build & any docs updated
- [x]  `spring-cloud-dataflow-server/docker-compose.yml` to include the `springcloud/spring-cloud-dataflow-grafana-influxdb` and `influxdb` images and to configure InfluxDB metrics by default. 
- [ ]  CI plan for DF should be updated to build & publish docker images
- [x]  k8s stuff references public image

### PR DESCRIPTION
**Pull request by tzolov**
- Build/push springcloud/spring-cloud-dataflow-grafana-prometheus and springcloud/spring-cloud-dataflow-grafana-influxdb to dockerhub (springcloud).
 - Use the pre-build grafana images inside all docker-compose.yml configurations.
 - Add grafana/influxdb metrics configuration enabled by default in the getting started Local docker-compose.yml.
 - Add getting started Doc section explaining the pre-configured Grafana/InfluxDB stream monitoring configuration.

Resolves #2900

### CODE REVIEW

**Comment by chrisjs** [comment]
i noticed (at the time of this writing):

* https://hub.docker.com/r/springcloud/spring-cloud-dataflow-grafana-prometheus/tags
  * the `latest` tag was published 12 minutes ago
  * the `2.0.0.BUILD-SNAPSHOT` tag was last updated 5 days ago

* https://hub.docker.com/r/springcloud/spring-cloud-dataflow-grafana-influxdb/tags
  * only has `2.0.0.BUILD-SNAPSHOT`

we should be publishing `latest` for each image which would be equiv to the latest dev version (ie: in this case `2.0.0.BUILD-SNAPSHOT`)

not sure if this is in flight or whatnot, but wanted to point it out as i just noticed it

File: spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc [hunk 1]
```diff
@@ -133,6 +133,14 @@ $ docker exec -it dataflow-server java -jar shell.jar
 
 Use of Spring Cloud Data Flow Shell is further described in <<shell,Shell>>.
 
+==== Spring Cloud Data Monitoring
+
+By default the Data Flow docker-compose configures Stream monitoring with InfluxDB and pre-build dashboards for Grafana.
```
Comment (chrisjs) [inline spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc]: `pre-build` i think should be `pre-built`

File: spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc [hunk 2]
```diff
@@ -133,6 +133,14 @@ $ docker exec -it dataflow-server java -jar shell.jar
 
 Use of Spring Cloud Data Flow Shell is further described in <<shell,Shell>>.
 
+==== Spring Cloud Data Monitoring
```
Comment (chrisjs) [inline spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc]: should `Spring Cloud Data Monitoring` be `Spring Cloud Data Flow Monitoring`?

File: spring-cloud-dataflow-server/docker-compose.yml [hunk 3]
```diff
@@ -36,7 +40,7 @@ services:
         do
           sleep 1;
         done;
-        wget -qO- 'http://dataflow-server:9393/apps' --post-data='uri=http://bit.ly/Darwin-SR2-stream-applications-kafka-maven&force=true';
+        wget -qO- 'http://dataflow-server:9393/apps' --post-data='uri=http://bit.ly/Einstein-BUILD-SNAPSHOT-stream-applications-kafka-maven&force=true';
```
Comment (chrisjs) [inline spring-cloud-dataflow-server/docker-compose.yml]: https://github.com/spring-cloud/spring-cloud-dataflow/pull/2924/files#diff-386250650dabfd12fc198e7a788707bcR712 is using `Einstein-GA`

File: src/grafana/influxdb/docker/docker-compose.yml [hunk 4]
```diff
@@ -1,24 +1,18 @@
...
 
-  # The grafana container uses the image resulting from the build
-  # of `./grafana` which simply packs some configuration into the
-  # form of an image.
-  #
-  # This could instead be a volume-mounted container.
+  # Grafana SCDF InfluxdDB pre-build image
   grafana:
-    build: './grafana'
+    # To use a locally build grafana image, comment out the 'image:' and uncomment 'build:' section. Then run 'docker-compose up -d --build`
```
Comment (chrisjs) [inline src/grafana/influxdb/docker/docker-compose.yml]: is there a need to use a locally built one vs the public image?

File: src/grafana/influxdb/docker/docker-compose.yml [hunk 5]
```diff
@@ -1,24 +1,18 @@
...
+    # build: './grafana'
+    image: springcloud/spring-cloud-dataflow-grafana-influxdb:${DATAFLOW_VERSION:?DATAFLOW_VERSION is not set! Use 'export DATAFLOW_VERSION=local-server-image-tag'}
```
Comment (chrisjs) [inline src/grafana/influxdb/docker/docker-compose.yml]: we don't have this check elsewhere, for consistency can it be removed or the others updated to do the same? IIRC if DF or Skipper versions aren't defined, there should be a message when doing `compose up` already?

File: src/grafana/prometheus/docker/grafana/Dockerfile [hunk 6]
```diff
@@ -1,4 +1,8 @@
-FROM grafana/grafana:5.4.2
+FROM grafana/grafana:5.4.3
 ADD ./provisioning /etc/grafana/provisioning
 ADD ./config.ini /etc/grafana/config.ini
 ADD ./dashboards /var/lib/grafana/dashboards
+ENV GF_INSTALL_PLUGINS digrich-bubblechart-panel,savantly-heatmap-panel,grafana-piechart-panel
+
```
Comment (chrisjs) [inline src/grafana/prometheus/docker/grafana/Dockerfile]: couple extra new lines here

**Comment by chrisjs** [comment]
sorry for any overlap, i think i commented on one of your commits rather than the PR itself. moved comments here..

File: spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc [hunk 7]
```diff
@@ -133,6 +133,14 @@ $ docker exec -it dataflow-server java -jar shell.jar
...
+
+By default the Data Flow docker-compose configures Stream monitoring with InfluxDB and pre-build dashboards for Grafana.
+
+For further instructions about the Data Flow monitoring follow the <<streams-monitoring-local-influx,Streams Monitoring InfluxDB >> section.
```
Comment (chrisjs) [inline spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc]: maybe - `instructions about Data Flow monitoring, follow the`

File: src/grafana/influxdb/docker/docker-compose.yml [hunk 8]
```diff
@@ -1,24 +1,18 @@
 version: '3.3'
 services:
 
-  # Prometheus uses the image resulting form the build
-  # of `./influxdb` which simply packs some configuration
-  # into the form of an image.
-  #
-  # It could instead, use volume-mounts.
+  # Prometheus uses the InfluxDB datasource
```
Comment (chrisjs) [inline src/grafana/influxdb/docker/docker-compose.yml]: would it be `Grafana uses the InfluxDB datasource`?

File: src/grafana/influxdb/docker/docker-compose.yml [hunk 9]
```diff
@@ -1,24 +1,18 @@
...
     container_name: 'influxdb'
     ports:
       - '8086:8086'
 
-  # The grafana container uses the image resulting from the build
-  # of `./grafana` which simply packs some configuration into the
-  # form of an image.
-  #
-  # This could instead be a volume-mounted container.
+  # Grafana SCDF InfluxdDB pre-build image
```
Comment (chrisjs) [inline src/grafana/influxdb/docker/docker-compose.yml]: `pre-built`

File: src/grafana/prometheus/docker/docker-compose.yml [hunk 10]
```diff
@@ -8,21 +8,19 @@ services:
...
-  # The grafana container uses the image resulting from the build
-  # of `./grafana` which simply packs some configuration into the
-  # form of an image.
-  #
-  # This could instead be a volume-mounted container.
-  # Use `docker exec -it grafana /bin/bash` to logging in the container
+  # Grafana SCDF Prometheus pre-build image:
   grafana:
-    build: './grafana'
+    # To use a locally build grafana image, comment out the 'image:' and uncomment 'build:' section. Then run 'docker-compose up -d --build`
```
Comment (chrisjs) [inline src/grafana/prometheus/docker/docker-compose.yml]: is there a need to use a locally built one vs the public SCDF custom image?

File: src/grafana/prometheus/docker/docker-compose.yml [hunk 11]
```diff
@@ -35,7 +33,7 @@ services:
   # Use `docker exec -it service-discovery /bin/sh` to logging the container
   # Project https://github.com/tzolov/spring-cloud-dataflow-prometheus-service-discovery
   service-discovery:
-    image: tzolov/spring-cloud-dataflow-prometheus-service-discovery:0.0.2
+    image: springcloud/spring-cloud-dataflow-prometheus-service-discovery:0.0.2
```
Comment (chrisjs) [inline src/grafana/prometheus/docker/docker-compose.yml]: wasn't it decided not to use this due to the dynamic port issue?

File: src/grafana/prometheus/docker/docker-compose.yml [hunk 12]
```diff
@@ -8,21 +8,19 @@ services:
   # It could instead, use volume-mounts.
   # Use `docker exec -it prometheus /bin/sh` to logging in the container
   prometheus:
-    build: './prometheus'
+    image: prom/prometheus:v2.7.1
     container_name: 'prometheus'
     volumes:
       - 'scdf-targets:/etc/prometheus/'
```
Comment (chrisjs) [inline src/grafana/prometheus/docker/docker-compose.yml]: see below about question on service discovery random port issue

File: spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc (hunk 1 above)
Comment (tzolov) [reply spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc]: +1

File: spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc (hunk 2 above)
Comment (tzolov) [reply spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc]: +1

File: spring-cloud-dataflow-server/docker-compose.yml (hunk 3 above)
Comment (tzolov) [reply spring-cloud-dataflow-server/docker-compose.yml]: +1

File: src/grafana/influxdb/docker/docker-compose.yml (hunk 4 above)
Comment (tzolov) [reply src/grafana/influxdb/docker/docker-compose.yml]: imo, this is handy for for testing/experimenting with uncommitted changes?

File: src/grafana/prometheus/docker/docker-compose.yml (hunk 11 above)
Comment (tzolov) [reply src/grafana/prometheus/docker/docker-compose.yml]: This is the only way to use the prometheus + grafana setup for SCDF local server. It expects though that the SCDF server is run on the host not via docker containers.

File: src/grafana/prometheus/docker/docker-compose.yml (hunk 10 above)
Comment (tzolov) [reply src/grafana/prometheus/docker/docker-compose.yml]: I've left those comments mostly for us, as a guideline how to debug  and test uncommitted changes.

File: src/grafana/influxdb/docker/docker-compose.yml (hunk 9 above)
Comment (tzolov) [reply src/grafana/influxdb/docker/docker-compose.yml]: +1

File: src/grafana/influxdb/docker/docker-compose.yml (hunk 8 above)
Comment (tzolov) [reply src/grafana/influxdb/docker/docker-compose.yml]: +1

File: spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc (hunk 7 above)
Comment (tzolov) [reply spring-cloud-dataflow-docs/src/main/asciidoc/getting-started-local.adoc]: +1

File: src/grafana/prometheus/docker/grafana/Dockerfile (hunk 6 above)
Comment (tzolov) [reply src/grafana/prometheus/docker/grafana/Dockerfile]: +1

File: src/grafana/influxdb/docker/docker-compose.yml (hunk 4 above)
Comment (chrisjs) [reply src/grafana/influxdb/docker/docker-compose.yml]: seems fairly straightforward to implement without the comments if needed. might be confusing to an end user as to why that's in there. if you feel strongly about keeping it, thats fine as well, maybe just make a note its for debugging or whatnot

File: src/grafana/prometheus/docker/docker-compose.yml (hunk 10 above)
Comment (chrisjs) [reply src/grafana/prometheus/docker/docker-compose.yml]: same reply as above

File: src/grafana/prometheus/docker/docker-compose.yml (hunk 11 above)
Comment (chrisjs) [reply src/grafana/prometheus/docker/docker-compose.yml]: right; maybe im misunderstanding something -- i thought on our last call prometheus on compose wasn't going to happen due to the host issue. we can chat offline if needed.

File: src/grafana/prometheus/docker/docker-compose.yml (hunk 12 above)
Comment (chrisjs) [reply src/grafana/prometheus/docker/docker-compose.yml]: in regards to the `scdf-targets` -- IIRC that was where the service discovery tool writes out app info and i thought that wasn't going to work in the compose setup. can talk about offline if im misunderstanding

File: src/grafana/influxdb/docker/docker-compose.yml (hunk 5 above)
Comment (chrisjs) [reply src/grafana/influxdb/docker/docker-compose.yml]: 👍

**Comment by tzolov** [comment]
> i noticed (at the time of this writing):
> 
> * https://hub.docker.com/r/springcloud/spring-cloud-dataflow-grafana-prometheus/tags
>   
>   * the `latest` tag was published 12 minutes ago
>   * the `2.0.0.BUILD-SNAPSHOT` tag was last updated 5 days ago
> * https://hub.docker.com/r/springcloud/spring-cloud-dataflow-grafana-influxdb/tags
>   
>   * only has `2.0.0.BUILD-SNAPSHOT`
> 
> we should be publishing `latest` for each image which would be equiv to the latest dev version (ie: in this case `2.0.0.BUILD-SNAPSHOT`)
> 
> not sure if this is in flight or whatnot, but wanted to point it out as i just noticed it

I was gonna test the CI plan after the merge

**Comment by chrisjs** [comment]
im getting the following in `spring-cloud-dataflow/spring-cloud-dataflow-server`:

```
$ DATAFLOW_VERSION=BUILD_SNAPSHOT SKIPPER_VERSION=2.0.0.BUILD_SNAPSHOT docker-compose up
ERROR: Invalid interpolation format for "image" option in service "dataflow-server": "springcloud/spring-cloud-dataflow-server:${DATAFLOW_VERSION:?DATAFLOW_VERSION is not set!}"
```

**Comment by sabbyanandan** [comment]
I was able to run all the things from scratch with the following two ways. 

1) Without env-vars:

```
➜ spring-cloud-dataflow/spring-cloud-dataflow-server ➤ b24b29ce|tzolov-gh-2900✓
± DATAFLOW_VERSION=2.0.0.BUILD-SNAPSHOT SKIPPER_VERSION=2.0.0.BUILD-SNAPSHOT docker-compose up     
```

2) With env-vars:

```
➜ spring-cloud-dataflow/spring-cloud-dataflow-server ➤ b24b29ce|tzolov-gh-2900✓
± export DATAFLOW_VERSION=2.0.0.BUILD-SNAPSHOT                                                                               ⏎ [3h]
➜ spring-cloud-dataflow/spring-cloud-dataflow-server ➤ b24b29ce|tzolov-gh-2900✓
± export SKIPPER_VERSION=2.0.0.BUILD-SNAPSHOT                                                                                ⏎ [3h]
➜ spring-cloud-dataflow/spring-cloud-dataflow-server ➤ b24b29ce|tzolov-gh-2900✓
± docker-compose up  
```

**Comment by chrisjs** [comment]
apparently my docker was too old... had to upgrade to latest and greatest

**Comment by chrisjs** [comment]
merged in fc5a72c424

**Pull request closed without merging**